examples/generative/pixelcnn.py

```diff
@@ -0,0 +1,184 @@
+"""
+
+import numpy as np
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+
+"""
+##Getting the Data
+"""
+
+# Model / data parameters
+num_classes = 10
+input_shape = (28, 28, 1)
+n_residual_blocks = 5
+# the data, split between train and test sets
+(x, _), (y, _) = keras.datasets.mnist.load_data()
+# Concatenate all of the images together
+data = np.concatenate((x, y), axis=0)
+# round all pixel values less than 33% of the max 256 value to 0
+# anything above this value gets rounded up to 1 so that all values are either
+# 0 or 1
+data = np.where(data < (0.33 * 256), 0, 1)
+data = data.astype(np.float32)
+
+"""
+## Create two classes for the requisite Layers for the model
+"""
+
+# The first layer is the PixelCNN layer. This layer simply
+# builds on the 2D convolutional layer, but includes masking.
+class PixelConvLayer(layers.Layer):
+    def __init__(self, mask_type, **kwargs):
+        super(PixelConvLayer, self).__init__()
+        self.mask_type = mask_type
+        self.conv = layers.Conv2D(**kwargs)
+
+    def build(self, input_shape):
+        # Build the conv2d layer to initialize kernel variables
+        self.conv.build(input_shape)
+        # Use the initialized kernel to create the mask
+        kernel_shape = self.conv.kernel.get_shape()
+        self.mask = np.zeros(shape=kernel_shape)
+        self.mask[: kernel_shape[0] // 2, ...] = 1.0
+        self.mask[kernel_shape[0] // 2, : kernel_shape[1] // 2, ...] = 1.0
+        if self.mask_type == "B":
+            self.mask[kernel_shape[0] // 2, kernel_shape[1] // 2, ...] = 1.0
+
+    def call(self, inputs):
+        self.conv.kernel.assign(self.conv.kernel * self.mask)
+        return self.conv(inputs)
+
+
+# Next, we build our residual block layer.
+# This is just a normal residual block, but base don the PixelConvLayer.
+class ResidualBlock(keras.layers.Layer):
+    def __init__(self, filters, **kwargs):
+        super(ResidualBlock, self).__init__(**kwargs)
+        self.activation = keras.layers.ReLU()
+        self.conv1 = keras.layers.Conv2D(
+            filters=filters, kernel_size=1, activation="relu"
+        )
+        self.pixel_conv = PixelConvLayer(
+            mask_type="B",
+            filters=filters // 2,
+            kernel_size=3,
+            activation="relu",
+            padding="same",
+        )
+        self.conv2 = keras.layers.Conv2D(
+            filters=filters, kernel_size=1, activation="relu"
+        )
+
+    def call(self, inputs):
+        x = self.activation(inputs)
+        x = self.conv1(x)
+        x = self.pixel_conv(x)
+        x = self.conv2(x)
+        return keras.layers.add([inputs, x])
+
+
+"""
+## Build the model based on the original paper
+"""
+
+inputs = keras.Input(shape=input_shape)
+x = PixelConvLayer(
+    mask_type="A", filters=128, kernel_size=7, padding="same", activation="relu"
+)(inputs)
+
+for _ in range(n_residual_blocks):
+    x = ResidualBlock(filters=128)(x)
```

Comment:
The initial input to this block is already relu-ed (by the layer at line 107), and then every subsequent input got relu-ed by the last `conv2` layer in the block. So it appears that the `ReLU` layer at the start of the block is not doing any work here. Unless I'm missing something?

Reply:
No you're right I must have been lazy and not thought of that connection when designing the residual block and model. I took out the one you mentioned before but I will say I wasn't sure whether it was necessary either as it was located after an addition layer. Model seems to do fine without it though.

Reply:
Both parts of the addition appear to be already relu-ed (one part is the input, which is always relu-ed before the block, and the other is the output of a conv-relu). So it should be safe to remove.

Reply:
Ok sounds good